src/platform/platform.dom.js

```diff
@@ -169,147 +164,109 @@ function throttled(fn, thisArg) {
 	};
 }
 
-function createDiv(cls) {
-	const el = document.createElement('div');
-	el.className = cls || '';
-	return el;
-}
-
-// Implementation based on https://github.com/marcj/css-element-queries
-function createResizer(domPlatform, handler) {
-	const maxSize = 1000000;
-
-	// NOTE(SB) Don't use innerHTML because it could be considered unsafe.
-	// https://github.com/chartjs/Chart.js/issues/5902
-	const resizer = createDiv(CSS_SIZE_MONITOR);
-	const expand = createDiv(CSS_SIZE_MONITOR + '-expand');
-	const shrink = createDiv(CSS_SIZE_MONITOR + '-shrink');
-
-	expand.appendChild(createDiv());
-	shrink.appendChild(createDiv());
-
-	resizer.appendChild(expand);
-	resizer.appendChild(shrink);
-	domPlatform._reset = function() {
-		expand.scrollLeft = maxSize;
-		expand.scrollTop = maxSize;
-		shrink.scrollLeft = maxSize;
-		shrink.scrollTop = maxSize;
-	};
-
-	const onScroll = function() {
-		domPlatform._reset();
-		handler();
-	};
-
-	addListener(expand, 'scroll', onScroll.bind(expand, 'expand'));
-	addListener(shrink, 'scroll', onScroll.bind(shrink, 'shrink'));
-
-	return resizer;
-}
-
-// https://davidwalsh.name/detect-node-insertion
-function watchForRender(node, handler) {
-	const expando = node[EXPANDO_KEY] || (node[EXPANDO_KEY] = {});
-	const proxy = expando.renderProxy = function(e) {
-		if (e.animationName === CSS_RENDER_ANIMATION) {
-			handler();
-		}
-	};
-
-	ANIMATION_START_EVENTS.forEach((type) => {
-		addListener(node, type, proxy);
+/**
+ * Watch for resize of `element`.
+ * Calling `fn` is limited to once per animation frame
+ * @param {Element} element - The element to monitor
+ * @param {function} fn - Callback function to call when resized
+ * @return {ResizeObserver=}
+ */
+function watchForResize(element, fn) {
+	const observer = new ResizeObserver(entries => {
+		const entry = entries[0];
+		helpers.requestAnimFrame.call(window, () => {
+			fn(entry.contentRect.width, entry.contentRect.height);
+		});
 	});
-
-	// #4737: Chrome might skip the CSS animation when the CSS_RENDER_MONITOR class
-	// is removed then added back immediately (same animation frame?). Accessing the
-	// `offsetParent` property will force a reflow and re-evaluate the CSS animation.
-	// https://gist.github.com/paulirish/5d52fb081b3570c81e3a#box-metrics
-	// https://github.com/chartjs/Chart.js/issues/4737
-	expando.reflow = !!node.offsetParent;
-
-	node.classList.add(CSS_RENDER_MONITOR);
+	observer.observe(element);
+	return observer;
 }
 
-function unwatchForRender(node) {
-	const expando = node[EXPANDO_KEY] || {};
-	const proxy = expando.renderProxy;
-
-	if (proxy) {
-		ANIMATION_START_EVENTS.forEach((type) => {
-			removeListener(node, type, proxy);
+/**
+ * Detect attachment of `element` or its direct `parent` to DOM
```

Comment:
Can this detect arbitrary elements up the DOM tree getting attached and unattached?

Reply:
No, we need to only detect if the canvas is detached from its container. ResizeObserver is bound to that and keeps functioning even when the tree is mutated.